microsite/blog/2023-12-13-backstage-and-optic.mdx

```diff
@@ -0,0 +1,208 @@
+
+One of our goals with the OpenAPI tooling project is to create a bootstrapping experience for existing plugins to easily lift-and-shift them onto OpenAPI tooling. I expect this command to be an integral part of that experience as it would be the core spec generation functionality.
+
+### Quick example using the Azure DevOps plugin
+
+Andre Wanlin asked me a few weeks ago to help him onboard the Azure DevOps plugin to the new OpenAPI tooling, with his main interest being the creation of an OpenAPI spec. [Here's](https://github.com/sennyeya/backstage/blob/dev-ops-openapi-spec/plugins/azure-devops-backend/src/schema/openapi.yaml) the spec that I ended up creating. If you were to clone this branch locally, you would find that there are a bunch of type errors. For example,
+
+```ts
+router.get("/projects", async (_req, res) => {
+  // no overload matches this call
+  const projects = await azureDevOpsApi.getProjects();
+  res.status(200).json(projects);
+});
+```
+
+and you'd see that there is no matching path for this endpoint in the spec or in the tests.
+
+Or,
+
+```ts
+router.get("/repo-builds/:projectName/:repoName", async (req, res) => {
+  const { projectName, repoName } = req.params;
+
+  const top = req.query.top ? Number(req.query.top) : DEFAULT_TOP; // Property 'top' does not exist on type
+
+  const gitRepository = await azureDevOpsApi.getRepoBuilds(
+    projectName,
+    repoName,
+    top
+  );
+
+  res.status(200).json(gitRepository);
+});
+```
+
+which, if you look at the spec,
+
+```yaml
+/repo-builds/{projectName}/{repoName}:
+  get:
+    parameters:
+      - in: path
+        name: projectName
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: repoName
+        required: true
+        schema:
+          type: string
+    responses:
+      "200":
+        description: 200 response
+        content:
+          application/json; charset=utf-8:
+            schema:
+              $ref: "#/components/schemas/GetBuildsProjectname200ResponseBody"
+```
+
+you'll see that there are no query parameters, and the tests don't include it,
+
+```ts
+describe('GET /repo-builds/:projectName/:repoName', () => {
+    it('fetches a list of repo builds', async () => {
+      ...
+      const response = await request(app)
+        .get('/repo-builds/myProject/myRepo')
+        .query({ top: '50' });
+
+      expect(azureDevOpsApi.getRepoBuilds).toHaveBeenCalledWith(
+        'myProject',
+        'myRepo',
+        50,
+      );
+      expect(response.status).toEqual(200);
+      expect(response.body).toEqual(repoBuilds);
+    });
+  });
+```
+
+By tying validation _and_ spec generation to unit tests, we expect unit test quality to increase, like adding the query param to the test above, as well as improving spec quality. If we can get the tooling right, it becomes a nice virtuous cycle of unit test improvements matched with spec improvements and the Backstage ecosystem wins.
```

Comment:
This is a really interesting idea -- I haven't heard many people expand on the second order effects of this in blogs before.